### ISSUE TITLE
Undocumented / incorrect requirement for import directives to be alphabetically sorted to trigger IDE0055

### ISSUE THREAD

**Issue by matt-edmondson** [dotnet/roslyn#77831]
**Version Used**:

.NET SDK Version: 9.0.201
SDK-Style Project
Microsoft Windows 11 Business 10.0.22631
Microsoft Visual Studio Enterprise 2022 (64-bit) - 17.13.5

**Steps to Reproduce**:

1. Have an SDK-Style C# project
2. Set `dotnet_separate_import_directive_groups = true` and `dotnet_diagnostic.IDE0055.severity = error` in `.editorconfig`
3. Add several usings in alphabetical order to a `.cs` file without separating line breaks
4. Observe an `IDE0055` formatting rule violation is triggered
5. Move one of the directives such that they are no longer all alphabetically sorted
6. Observe that the `IDE0055` formatting rule violation is no longer triggered

```csharp
# Triggers IDE0055
namespace ktsu.IDE0055Test;

using Azure.Storage.Blobs;
using Azure.Storage.Sas;
using NuGet.Versioning;
using System.Diagnostics;
```

```csharp
# Does not trigger IDE0055
namespace ktsu.IDE0055Test;

using Azure.Storage.Blobs;
using Azure.Storage.Sas;
using System.Diagnostics;
using NuGet.Versioning;
```

There is a comment in the analyzer rule that alludes to this as if it were intended behavior:

https://github.com/dotnet/roslyn/blob/a35ddc34f2bfc272d0ad9519aeaf75c5bd6a4d9f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs#L225

```csharp
        // if the user is separating using-groups, and we're between two usings, and these
        // usings *should* be separated, then do so (if the usings were already properly
        // sorted).
```

However the documentation for configuring the rule makes no mention of this requirement:

https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/dotnet-formatting-options#dotnet_separate_import_directive_groups

**Diagnostic Id**: IDE0055: Fix formatting

**Expected Behavior**:

`UsingsAndExternAliasesOrganizer.NeedsGrouping()` indicates that two namespaces should belong to the same grouping if the first token of each is different.

https://github.com/dotnet/roslyn/blob/a35ddc34f2bfc272d0ad9519aeaf75c5bd6a4d9f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/UsingsAndExternAliasesOrganizer.cs#L66

The rule should not enforce alphabetical sorting, or the requirement should be documented, and if the sorting is intentional it should properly respect `dotnet_sort_system_directives_first` instead of allowing either completely sorted or sorted with `System` namespaces first.

In summary the rule should:

- Not query for alphabetical sorting at all

or

- Query for alphabetical sorting in a way that respects the `dotnet_sort_system_directives_first` configuration, and be documented as such

**Actual Behavior**:

`TokenBasedFormattingRule.AdjustNewLinesAfterSemicolonToken()` will only enforce separating newlines between grouped usings if all of the usings are sorted either alphabetically, or alphabetically with the `System` group first, regardless of the configuration of `dotnet_sort_system_directives_first`.

https://github.com/dotnet/roslyn/blob/a35ddc34f2bfc272d0ad9519aeaf75c5bd6a4d9f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/TokenBasedFormattingRule.cs#L225

```csharp
            if (usings.IsSorted(UsingsAndExternAliasesDirectiveComparer.SystemFirstInstance) ||
                usings.IsSorted(UsingsAndExternAliasesDirectiveComparer.NormalInstance))
```

This check also does not match the documentation for `dotnet_sort_system_directives_first` in that it states:

```
Sort System.* using directives alphabetically, and place them before other using directives.
```

But actually checks for:

```
Sort System.* using directives alphabetically, and place them before other alphabetically sorted using directives.
```

**Comment by sharwell** [comment]
IDE0055 only reports cases where whitespace can be added to fix a violation, or whitespace can be removed to fix a violation. It's not clear from the above how we would expect IDE0055 to behave in this case.

**Comment by JoeRobich** [comment]
I think having a separate analyzer/codefix for organized imports would make sense.

**Comment by matt-edmondson** [comment]
> IDE0055 only reports cases where whitespace can be added to fix a violation, or whitespace can be removed to fix a violation. It's not clear from the above how we would expect IDE0055 to behave in this case.

I would expect that it uses the same logic as `UsingsAndExternAliasesOrganizer.NeedsGrouping()` and inserts a newline when the first token of the previous and current directive names do not match.

In the case that the directives are not alphabetically sorted this could result in rather sparse groupings, which is not ideal, but maybe exposes that more is needed to properly facilitate this rule as it is described.

My opinion is that the sorting is an additional responsibility that should either be shifted into its own diagnostic as @JoeRobich suggested, or formalize the requirement by expanding the documentation to describe it and make it correctly respect `dotnet_sort_system_directives_first`.
